### PR TITLE
refine rotated object detector

### PR DESCRIPTION
**Pull request by Asthestarsfalll**
### PR types

Bug fixes
### PR changes

Others
### Description

修复#100中旋转目标检测的问题，添加PPYOLOE—R

### CODE REVIEW

File: paddlers/tasks/object_detector.py [hunk 1]
```diff
@@ -303,6 +305,10 @@ def train(self,
             early_stop_patience (int, optional): Early stop patience. Defaults to 5.
             use_vdl(bool, optional): Whether to use VisualDL to monitor the training 
                 process. Defaults to True.
+            clip_grad_by_norm (float, optional): Maximum global norm for gradient clipping. 
+                Default: None.
+            reg_coeff (float, optional): Coefficient for L2 weight decay regularization.
```
Comment (Bobholamovic) [inline paddlers/tasks/object_detector.py:326]: 需同步更新文档

**Review by Bobholamovic — COMMENTED** [review]
这个设计可以的，我觉得ok。
另外在这个PR中也辛苦修复一下backbone的bug吧。

File: paddlers/tasks/object_detector.py (hunk 1 above)
Comment (Asthestarsfalll) [reply paddlers/tasks/object_detector.py:326]: 已更新

**Comment by Asthestarsfalll** [comment]
修复了DarkNet53的问题，同时重构了一下backbone的检查。

File: tutorials/train/object_detection/ppyoloe_r.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+# 数据集图像目录
+IMAGE_DIR = "./data/dota/images"
+# 实验目录，保存输出的模型权重和结果
+EXP_DIR = "./output/ppyoloe_r/"
+
+IMAGE_SIZE = [1024, 1024]
+
+# 下载和解压SAR影像舰船检测数据集
+pdrs.utils.download_and_decompress(
+    "https://paddlers.bj.bcebos.com/datasets/dota.zip", path="./data/")
```
Comment (Asthestarsfalll) [inline tutorials/train/object_detection/ppyoloe_r.py:22]: 另外这里的数据集似乎需要添加一下

**Comment by Asthestarsfalll** [comment]
ppyoloe-r训练结果，使用前600多样本

File: paddlers/tasks/base.py [hunk 3]
```diff
@@ -386,6 +386,8 @@ def train_loop(self,
             step_time_tic = time.time()
 
             for step, data in enumerate(self.train_data_loader()):
+                if self.model_name in ['PicoDet', 'PPYOLOE_R']:
```
Comment (Bobholamovic) [inline paddlers/tasks/base.py:391]: 这里让基类依赖于子类感觉不是很优雅。如果没有更好的解决方案的话，可以加个`# TODO: Refactor` 注释先标记一下。

Comment (Bobholamovic) [reply paddlers/tasks/base.py:391]: 另外这里是为什么需要进行这个操作呀？

File: tutorials/train/object_detection/ppyolo.py [hunk 4]
```diff
@@ -78,6 +78,7 @@
     save_interval_epochs=5,
     # 每多少次迭代记录一次日志
     log_interval_steps=4,
+    metric='rbox',
```
Comment (Bobholamovic) [inline tutorials/train/object_detection/ppyolo.py]: 这里是为啥需要修改哦？

File: tutorials/train/object_detection/ppyoloe_r.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+# 对于旋转目标检测任务，需要安装自定义外部算子库，安装方式如下：
+# cd paddlers/models/ppdet/ext_op
+# python setup.py install
+
+# 定义训练和验证时使用的数据变换（数据增强、预处理等）
+# 使用Compose组合多种变换方式。Compose中包含的变换将按顺序串行执行
+# API说明：https://github.com/PaddlePaddle/PaddleRS/blob/develop/docs/apis/data.md
+train_transforms = T.Compose([
```
Comment (Bobholamovic) [inline tutorials/train/object_detection/ppyoloe_r.py]: 建议可以去除`T.Compose`以及`T.BatchCompose`，和其它tutorials保持相同风格。

**Comment by Asthestarsfalll** [comment]
ppyoloe_r训练结果，这里保存路径忘改了

File: paddlers/tasks/object_detector.py
Comment (Bobholamovic) [inline paddlers/tasks/object_detector.py:1]: 我发现上个PR对batch transforms的重构好像有一定问题，原先的行为是用户指定的batch transforms会和内置的batch transforms混合，现在好像变成了要么用custom要么用default了……可能需要修改一下保证行为不变。

**Review by Bobholamovic — CHANGES_REQUESTED** [review]
目前除了旋转框检测模型的脚本，其它的检测模型的tutorial好像都跑不起来了……辛苦在修改代码后验证一下原有模型是否能够跑通，咱们要确保新增模型不会破坏已有模型哈~

另外我提了一个commit修复了一些小问题。

File: tutorials/train/object_detection/ppyolo.py (hunk 4 above)
Comment (Asthestarsfalll) [reply tutorials/train/object_detection/ppyolo.py]: 加错文件了...

File: tutorials/train/object_detection/ppyoloe_r.py (hunk 5 above)
Comment (Asthestarsfalll) [reply tutorials/train/object_detection/ppyoloe_r.py]: 了解

File: paddlers/tasks/base.py (hunk 3 above)
Comment (Asthestarsfalll) [reply paddlers/tasks/base.py:391]: 这里是因为这两个网络需要在不同的训练阶段使用不同的assigner。
https://github.com/PaddlePaddle/PaddleRS/blob/dfd50caf860202336e47f6bdcc7dd1863f2f85aa/paddlers/models/ppdet/modeling/heads/pico_head.py#L653-L674

https://github.com/PaddlePaddle/PaddleRS/blob/dfd50caf860202336e47f6bdcc7dd1863f2f85aa/paddlers/models/ppdet/modeling/heads/ppyoloe_r_head.py#L260-L283

注意到dataset有`set_epoch`方法，可以考虑直接在__get_item__里面设置

File: paddlers/tasks/object_detector.py
Comment (Asthestarsfalll) [reply paddlers/tasks/object_detector.py:1]: 如果是混合的话就不能和ppdet中进行对齐了，而且这是否是用户已知的行为，个人认为未指定就使用默认比较好，混合似乎会对用户造成一定困扰（如果是我的话）

File: paddlers/tasks/object_detector.py [hunk 6]
```diff
@@ -2187,5 +2158,180 @@ def _get_test_inputs(self, image_shape):
...
+                assigner=assigner,
+                nms=nms)
+            params.update({
+                'backbone': backbone,
+                'neck': neck,
+                'yolo_head': yolo_head,
+                'post_process': post_process
+            })
+        # NOTE: call BaseDetector.__init__ instead of YOLOv3.__init__
+        super(YOLOv3, self).__init__(
```
Comment (Asthestarsfalll) [inline paddlers/tasks/object_detector.py:2381]: 之前的方法似乎有些多余，因此还是修改为类似其他继承YOLO3的方法。

File: paddlers/utils/checkpoint.py [hunk 7]
```diff
@@ -500,6 +512,13 @@ def get_pretrain_weights(flag, class_name, save_dir, backbone_name=None):
     return fname
 
 
+def is_without_backbone(state_dict):
+    for k in state_dict.keys():
+        if k.startswith('backbone'):
+            return False
+    return True
+
+
```
Comment (Asthestarsfalll) [inline paddlers/utils/checkpoint.py]: 这里的改动是因为CSPResNet的权重都少了一个`backbone.`

File: paddlers/tasks/object_detector.py
Comment (Bobholamovic) [reply paddlers/tasks/object_detector.py:1]: 这个地方原先确实有些问题，在用户不知情的情况下加入了和修改了一些东西，我能想到的唯一的好处可能就是：减轻用户配置的负担（用户不用手动加这些可能是『必须要有』的变换，也不用关心`collate_batch`应该设置为True还是False）。从编程的角度来说，我不喜欢这样『隐式』的行为，我十分赞成变量的名字和功能应该匹配。

不过，在我们修改了代码的行为之后，原有的tutorial脚本跑不起来了（因为原有的脚本基于这种『混合』的行为），我想这应该是一件更大的事——代码好看了，但是用不了。从重构的角度来说，我想我们不算成功，因为重构不应该改变代码外在的行为。为了解决这个『更严重』的问题，我有两个方案，我们可以讨论一下：
1. 保持优雅的代码设计，即，要么使用用户定义的batch_transforms，要么使用默认的。这种方式需要我们修改tutorials脚本中的示例代码，让它们能够基于新的API跑通。风险在于我们可能会发现，tutorials脚本变得异常复杂，对用户极不友好，而这违背了PaddleRS产品的原则，之后我们又要进一步思考如何隐藏一些细节，让用户使用起来『更简单』。
2. 偷个懒，改回原先的逻辑，让两种transforms按照原有的方式混合，hard-code那些collate_batch，以确保tutorials脚本不必变动就能跑出原来的效果。这样的缺点在于我们不得不接受重构前的那个丑陋的设计，并且这个设计造成的负面影响将继续存在。

File: paddlers/tasks/base.py (hunk 3 above)
Comment (Bobholamovic) [reply paddlers/tasks/base.py:391]: 如果需要在这里进行的话也是OK的，可以在注释里稍微标注一下原因。

File: paddlers/tasks/object_detector.py (hunk 6 above)
Comment (Bobholamovic) [reply paddlers/tasks/object_detector.py:2381]: OK

File: paddlers/utils/checkpoint.py (hunk 7 above)
Comment (Bobholamovic) [reply paddlers/utils/checkpoint.py]: `BaseModel.initialize_net`有一个`is_backbone_weights`参数，是不是可以在`BaseDetector`里改造一下？

Comment (Asthestarsfalll) [reply paddlers/utils/checkpoint.py]: 已修改

**Comment by Asthestarsfalll** [comment]
不知道为什么这里的batch_transforms就是为None, COCODetDataset是正常的

https://github.com/PaddlePaddle/PaddleRS/blob/98af92f90079d442b2aa218500399caa684b71d2/tutorials/train/object_detection/faster_rcnn.py#L56-L62

File: paddlers/tasks/object_detector.py [hunk 8]
```diff
@@ -2239,6 +2260,26 @@ def __init__(self,
         self.anchors = anchors
         self.anchor_masks = anchor_masks
 
+    def _default_batch_transforms(self, mode='train'):
+        if mode == 'train':
+            batch_transforms = [
+                BatchNormalizeImage(
```
Comment (Bobholamovic) [inline paddlers/tasks/object_detector.py]: `normalize`的部分不建议放在`_default_batch_transforms`中，因为定制归一化参数是比较常见的需求~

Comment (Asthestarsfalll) [reply paddlers/tasks/object_detector.py]: 了解

File: paddlers/tasks/object_detector.py
Comment (Asthestarsfalll) [reply paddlers/tasks/object_detector.py:1]: 此处按照第二种方式修改

**Comment by Asthestarsfalll** [comment]
教程脚本测试结果：
ppyoloe_r:

fcosr:

ppyolo_tiny:

yolov3:

ppyolov2:

ppyolo：始终为0
faster rcnn：loss为nan

**Comment by Asthestarsfalll** [comment]
faster rcnn:

File: paddlers/tasks/object_detector.py [hunk 9]
```diff
@@ -454,6 +475,26 @@ def _collate_fn(batch):
     def _default_batch_transforms(self, mode):
         raise NotImplementedError
 
+    def _filter_batch_transforms(self, defaults, targets):
+        if targets is None:
+            return defaults
+        target_types = [type(i) for i in targets]
+        filtered = [i for i in defaults if type(i) not in target_types]
+        return filtered
```
Comment (Asthestarsfalll) [inline paddlers/tasks/object_detector.py:484]: 这里用于过滤default batch tansforms中与用户自定义的batch transforms的重复项

Comment (Bobholamovic) [reply paddlers/tasks/object_detector.py:484]: OK
